janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java

```diff
@@ -114,8 +115,33 @@ public void setup(final Map<String,Object> config) {
         }
 
         final String defaultUser = config.get(CONFIG_DEFAULT_USER).toString();
-        if (credentialStore.findUser(defaultUser) == null) {
-            credentialStore.createUser(defaultUser, config.get(CONFIG_DEFAULT_PASSWORD).toString());
+        if (!userExists(defaultUser)) {
+            createUser(defaultUser, config.get(CONFIG_DEFAULT_PASSWORD).toString());
+        }
+    }
+
+    protected Vertex findUser(final String username) {
+        credentials.tx().rollback();
+        final GraphTraversal<Vertex,Vertex> t = credentials.users(username);
+        final Vertex v = t.hasNext() ? t.next() : null;
+        if (t.hasNext()) throw new IllegalStateException(String.format("Multiple users with username %s", username));
+        return v;
+    }
+
+    private boolean userExists(final String username) {
+        credentials.tx().rollback();
```

Comment:
Same here

Reply:
Same answer as above basically. Instead of `userExists()` the code previously did this to perform the existence check:
```java
credentialStore.findUser(defaultUser) == null
```
Since `findUser()` had the rollback, I also added it here.

Reply:
@FlorianHockmann should we throw an exception in this method if multiple users exist?

Reply:
`userExists()` is currently only called in line 119 for the `defaultUser`:

```java
final String defaultUser = config.get(CONFIG_DEFAULT_USER).toString();
if (!userExists(defaultUser)) {
    createUser(defaultUser, config.get(CONFIG_DEFAULT_PASSWORD).toString());
}
```

This code will be executed on every start of JanusGraph and it's expected that the `defaultUser` exists after the first start of JanusGraph. So, at least in this case we should not throw an exception.

Reply:
@FlorianHockmann I am just interested in the flow when there multiple default users exist. Is it a normal case when multiple same users exist? I didn't check but I assume that this is a possible scenario when we create multiple default users. Should we remove unnecessary users or they will not disturb us?

Reply:
I'm not sure how multiple default users should exist. You specify the default user in the config like this:
```
config: {
     defaultUsername: user,
     defaultPassword: password,
     credentialsDb: conf/janusgraph-credentials-server.properties
}
```

You could of course change the `defaultUsername` and start JanusGraph again which would then create another _default_ user, but I don't see how we could prevent that as JanusGraph won't know any more that the already existing user _user_ was the default user before. We would have to either remove **all** existing users (sounds like a very bad idea to me) or mark the default user somehow so we can check whether such a default user already exists and then delete all old ones. If we want to follow that path, then we should do it in a separate PR in my opinion.
I'm however not sure in general how the credential graph should be managed right now. We should probably at something for that to our docs.

Reply:
@FlorianHockmann sure, I don't tell to do it in this PR. Possibly we should discuss it somewhere else. I just thought about consistency and race conditions when we creating a default user.
1. We could use eventual consistent storage backend. That is why `userExists(defaultUser)` doesn't guarantee that the user exists or not exists. For example, if we use Cassandra with small consistency level or with multi data center setup and we start multiple instances together then this method isn't sufficient (unless it uses some locking mechanism).
2. Again, even if we use one DC with big consistency level there is a small time gap between `userExists(...)` and `createUser(...)` which may be enough to execute `userExists(...)` two times and return `false`.

Reply:
Ah, now I see what you mean. I really didn't think about race conditions due to eventual consistency at all here :-/

I think an issue makes sense for this case. Do you want to create one?

Reply:
Sure. I've created it here #1434